agent/src/protocols/udev/discovery_impl.rs

```diff
@@ -152,6 +152,7 @@ fn find_devices(
             if let Some(devnode) = get_devnode(&device) {
                 Some(devnode.to_str().unwrap().to_string())
             } else {
+                trace!("find_devices - ignoring device with devpath {:?} due to having no devnode", get_devpath(&device));
```

Comment:
Not all devices have devnodes. For instance, the udev rule `SUBSYSTEM=="usb"` finds 9 devpaths on my machine but only 4 of those have devnodes, so while these [testing udev rule](https://github.com/deislabs/akri/blob/main/docs/udev-configuration.md#testing-a-udev-rule) steps show 9 devices, Akri will only create instances for the ones that have devnodes. Should this be changed to using devpath like so:
```
    let device_devpaths: Vec<String> = final_devices
        .into_iter()
        .filter_map(|device| {
            Some(get_devpath(&device).to_str().unwrap().to_string())
        })
        .collect();
```
The udev video broker expects a devnode, so it would need a modification.
It seems that selecting for devnodes makes sense for finding devices, but it is harder to test. I am not quite sure why some items in udev have devpaths but not devnodes. Anyone have an understanding of what qualifies a device for having a devnode vs not?

Reply:
Curious, are the 9 devices all leaf devices? Are the remaining 5 not exposed under /dev/. but also not under /sys?

Reply:
They are all exposed under `/sys`. Everything in the Linux Device Filesystem has a syspath; however, not everything has a dev node. For example, all syspaths discovered by `SUBSYSTEM=="usb"`:
/sys/devices/pci0000:00/0000:00:14.0/usb1
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-0:1.0
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/hidraw/hidraw0
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/event3
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/input6::capslock
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/input6::numlock
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/input6::scrolllock
/sys/devices/pci0000:00/0000:00:14.0/usb2
/sys/devices/pci0000:00/0000:00:14.0/usb2/2-0:1.0

However, only the following have devnodes:
/dev/bus/usb/001/001
/dev/bus/usb/001/002
/dev/hidraw0
/dev/input/event3
/dev/bus/usb/002/001

Reply:
This is a nice write up: http://reactivated.net/writing_udev_rules.html

Anyways, I think there might be several sources for /dev nodes. 

/etc/udev/rules.d + /lib/udev/rules.d and /usr/lib/rules.d have rules that contain `NAME=` and/or `SYMLINK+=` clauses, which create symlinks  under /dev.

I recall that kernel drivers can also directly create nodes under /dev. Looking at the rules above on my box, I did not see the rules that would explain the dev nodes, so perhaps that is coming out of drivers (though perhaps there is some % substitution in play?).

From the link above, wonder if this would help figure out origin of the 5 devices in /dev:
```
udevtest
If you know the top-level device path in sysfs, you can use udevtest to show the actions which udev would take. This may help you debug your rules. For example, assuming you want to debug a rule which acts on /sys/class/sound/dsp:

# udevtest /class/sound/dsp
main: looking at device '/class/sound/dsp' from subsystem 'sound'
udev_rules_get_name: add symlink 'dsp'
udev_rules_get_name: rule applied, 'dsp' becomes 'sound/dsp'
udev_device_event: device '/class/sound/dsp' already known, remove possible symlinks
udev_node_add: creating device node '/dev/sound/dsp', major = '14', minor = '3', mode = '0660', uid = '0', gid = '18'
udev_node_add: creating symlink '/dev/dsp' to 'sound/dsp'
```
Note the usage is `udevadm test`, not `udevtest`.

In my case:
```
GROUP 104 /lib/udev/rules.d/50-udev-default.rules:29
IMPORT builtin 'hwdb' /lib/udev/rules.d/60-evdev.rules:8
IMPORT builtin 'hwdb' returned non-zero
IMPORT builtin 'hwdb' /lib/udev/rules.d/60-evdev.rules:17
IMPORT builtin 'hwdb' returned non-zero
IMPORT builtin 'hwdb' /lib/udev/rules.d/60-evdev.rules:21
IMPORT builtin 'hwdb' returned non-zero
IMPORT builtin 'input_id' /lib/udev/rules.d/60-input-id.rules:5
capabilities/ev raw kernel attribute: 120013
capabilities/abs raw kernel attribute: 0
capabilities/rel raw kernel attribute: 0
capabilities/key raw kernel attribute: 1000000000007 ff9f207ac14057ff febeffdfffefffff fffffffffffffffe
properties raw kernel attribute: 0
test_key: checking bit block 0 for any keys; found=1
test_key: checking bit block 64 for any keys; found=1
test_key: checking bit block 128 for any keys; found=1
test_key: checking bit block 192 for any keys; found=1
IMPORT builtin 'hwdb' /lib/udev/rules.d/60-input-id.rules:6
IMPORT builtin 'hwdb' returned non-zero
IMPORT builtin 'usb_id' /lib/udev/rules.d/60-persistent-input.rules:11
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0: if_class 3 protocol 0
LINK 'input/by-id/usb-Dell_Dell_USB_Keyboard-event-kbd' /lib/udev/rules.d/60-persistent-input.rules:28
IMPORT builtin 'path_id' /lib/udev/rules.d/60-persistent-input.rules:35
LINK 'input/by-path/pci-0000:00:14.0-usb-0:14:1.0-event-kbd' /lib/udev/rules.d/60-persistent-input.rules:37
PROGRAM 'libinput-device-group /sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/event3' /lib/udev/rules.d/80-libinput-device-groups.rules:7
starting 'libinput-device-group /sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/event3'
'libinput-device-group /sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/event3'(out) '3/413c/2003:usb-0000:00:14.0-14'
Process 'libinput-device-group /sys/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/event3' succeeded.
IMPORT builtin 'hwdb' /lib/udev/rules.d/90-libinput-model-quirks.rules:42
IMPORT builtin 'hwdb' returned non-zero
IMPORT builtin 'hwdb' /lib/udev/rules.d/90-libinput-model-quirks.rules:46
IMPORT builtin 'hwdb' returned non-zero
IMPORT builtin 'hwdb' /lib/udev/rules.d/90-libinput-model-quirks.rules:50
IMPORT builtin 'hwdb' returned non-zero
handling device node '/dev/input/event3', devnum=c13:67, mode=0660, uid=0, gid=104
preserve permissions /dev/input/event3, 020660, uid=0, gid=104
preserve already existing symlink '/dev/char/13:67' to '../input/event3'
found 'c13:67' claiming '/run/udev/links/\x2finput\x2fby-id\x2fusb-Dell_Dell_USB_Keyboard-event-kbd'
creating link '/dev/input/by-id/usb-Dell_Dell_USB_Keyboard-event-kbd' to '/dev/input/event3'
preserve already existing symlink '/dev/input/by-id/usb-Dell_Dell_USB_Keyboard-event-kbd' to '../event3'
found 'c13:67' claiming '/run/udev/links/\x2finput\x2fby-path\x2fpci-0000:00:14.0-usb-0:14:1.0-event-kbd'
creating link '/dev/input/by-path/pci-0000:00:14.0-usb-0:14:1.0-event-kbd' to '/dev/input/event3'
preserve already existing symlink '/dev/input/by-path/pci-0000:00:14.0-usb-0:14:1.0-event-kbd' to '../event3'
.INPUT_CLASS=kbd
.MM_USBIFNUM=00
ACTION=add
BACKSPACE=guess
DEVLINKS=/dev/input/by-id/usb-Dell_Dell_USB_Keyboard-event-kbd /dev/input/by-path/pci-0000:00:14.0-usb-0:14:1.0-event-kbd
DEVNAME=/dev/input/event3
DEVPATH=/devices/pci0000:00/0000:00:14.0/usb1/1-14/1-14:1.0/0003:413C:2003.0001/input/input6/event3
ID_BUS=usb
ID_INPUT=1
ID_INPUT_KEY=1
ID_INPUT_KEYBOARD=1
ID_MODEL=Dell_USB_Keyboard
ID_MODEL_ENC=Dell\x20USB\x20Keyboard
ID_MODEL_ID=2003
ID_PATH=pci-0000:00:14.0-usb-0:14:1.0
ID_PATH_TAG=pci-0000_00_14_0-usb-0_14_1_0
ID_REVISION=0306
ID_SERIAL=Dell_Dell_USB_Keyboard
ID_TYPE=hid
ID_USB_DRIVER=usbhid
ID_USB_INTERFACES=:030101:
ID_USB_INTERFACE_NUM=00
ID_VENDOR=Dell
ID_VENDOR_ENC=Dell
ID_VENDOR_ID=413c
LIBINPUT_DEVICE_GROUP=3/413c/2003:usb-0000:00:14.0-14
MAJOR=13
MINOR=67
SUBSYSTEM=input
TAGS=:power-switch:akri_tag:
USEC_INITIALIZED=5419248
XKBLAYOUT=us
XKBMODEL=pc105
XKBOPTIONS=
XKBVARIANT=
Unload module index
Unloaded link configuration context.
```
Seems that my case might be your case too based on the presence of the TAG :).

Perhaps this file is responsible? `/lib/udev/rules.d/60-input-id.rules`

Coming back to your original question, would it make sense to craft a rule with more filters, not just `SUBSYSTEM=="usb"`, but perhaps adding a clause for video, or in the example above, input?

Reply:
To clarify: Are you saying that we should use syspath instead of devpath and leave it on the operator to create more specific rules to narrow down the device list? Or that most devices that should be discovered will have devnodes as the drivers for those devices will most likely define rules with NAME= or SYMLINK+= ?

Reply:
Also, I just removed the tag. I am playing with rules in `/etc/udev/rules.d/90-akri.rules`.

Reply:
My hunch is to switch to using `syspath` and modify the udev video broker

Reply:
I was not meaning to imply that (I meant to say that either drivers do create the /dev/* nodes or there is udev rule for those). For /dev/input/event3 specifically, there was some udev rule as per the listing above. Also looking specifically under /sys/devices for /dev/input/event3, I do not see a file that could be mounted into container. So I think we need to pick the files from other /dev.